### PR TITLE
WINC-751: Upgrade sdk-operator to 1.32.0

### PR DESCRIPTION
**Pull request by alinaryan**
Updates the OSDK version to latest v1.32.0.

- Added updates required for Golang/v3 based operators following https://sdk.operatorframework.io/docs/upgrading-sdk-version/ starting from version 1.16.0 to latest 1.32.0.

- Updates the OSDK version in our pre-release.sh script and the hacking script.

- Skipped anything that did not apply to our operator (helm/ansible/non-Go-v3 based operators)

### CODE REVIEW

**Review by aravindhp — CHANGES_REQUESTED** [review]
I did not do a full review but noticed that you missed updating [get_operator_sdk()](https://github.com/openshift/windows-machine-config-operator/blob/2cd002bbc30f981685e1bffd1da749a81e9ecfa2/hack/common.sh#L137). Please fix and kick off an e2e that exercises that function.

File: hack/pre-release.sh [hunk 1]
```diff
@@ -227,7 +227,7 @@ github_update() {
 
   # remove make bundle artifacts
   sed -i 's/REPLACE_IMAGE:latest/REPLACE_IMAGE/' bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
-  sed -i 's/operator-sdk-v1.14.0+git/operator-sdk-v1.15.0+git/' bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
+  sed -i 's/operator-sdk-v1.14.0+git/operator-sdk-v1.32.0+git/' bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
```
Comment (aravindhp) [inline hack/pre-release.sh:230]: Please use the `OPERATOR_SDK_VERSION` variable here so that we don't need to update this line for future SKD version changes

**Review by aravindhp — CHANGES_REQUESTED** [review]

**Comment by jrvaldes** [comment]
@alinaryan  consider mentioning explicitly the version you are updating to i.e. `1.32` in the PR title instead of "latest"

File: Makefile [hunk 2]
```diff
@@ -47,6 +47,7 @@ else
 GOBIN=$(shell go env GOBIN)
 endif
 
+.PHONY: all
```
Comment (jrvaldes) [inline Makefile:50]: Are these changes required for the sdk-operator upgrade? If so, please mention why in the commit message; otherwise, I recommend moving them to a separate PR.

File: Makefile [hunk 3]
```diff
@@ -147,8 +147,8 @@ CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
 ENVTEST ?= $(LOCALBIN)/setup-envtest
 
 ## Tool Versions
-KUSTOMIZE_VERSION ?= v3.8.7
-CONTROLLER_TOOLS_VERSION ?= v0.8.0
+KUSTOMIZE_VERSION ?= v4.5.5
+CONTROLLER_TOOLS_VERSION ?= v0.10.0
```
Comment (saifshaikh48) [inline Makefile:151]: since these were introduced in the last commit, I recommend squashing this commit

File: Makefile [hunk 4]
```diff
@@ -154,17 +154,17 @@ KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/k
 .PHONY: kustomize
 kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
 $(KUSTOMIZE): $(LOCALBIN)
-	curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN)
+	test -s $(LOCALBIN)/kustomize || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
```
Comment (saifshaikh48) [inline Makefile:157]: Same as previous comment, I recommend squashing this commit into `[build] Replace go-get with go-install`

**Review by saifshaikh48 — COMMENTED** [review]
Thanks for doing this, just a few minor comments

File: Makefile (hunk 2 above)
Comment (alinaryan) [reply Makefile:50]: Yes they are regarded as part of the OSDK upgrade. See https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.16.0/#add-phony-targets-to-makefile

Comment (aravindhp) [reply Makefile:50]: Please add such links to your commit message. This will explain WHY you are making the change.

Comment (alinaryan) [reply Makefile:50]: Will add links to this and the rest of the commits

File: config/default/manager_auth_proxy_patch.yaml [hunk 5]
```diff
@@ -15,7 +15,7 @@ spec:
         - "--secure-listen-address=0.0.0.0:8443"
         - "--upstream=http://127.0.0.1:8080/"
         - "--logtostderr=true"
-        - "--v=10"
+        - "--v=0"
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply config/default/manager_auth_proxy_patch.yaml:18]: No, but I'll add the link to the doc in commit message

File: Makefile [hunk 6]
```diff
@@ -173,6 +173,16 @@ OPM = $(shell which opm)
 endif
 endif
 BUNDLE_IMGS ?= $(BUNDLE_IMG)
+# BUNDLE_GEN_FLAGS are the flags passed to the operator-sdk generate bundle command
+BUNDLE_GEN_FLAGS ?= -q --overwrite --version $(WMCO_VERSION) $(BUNDLE_METADATA_OPTS)
+
+# USE_IMAGE_DIGESTS defines if images are resolved via tags or digests
+# You can enable this value if you would like to use SHA Based Digests
+# To enable set flag to true
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:193]: Added

File: cmd/operator/main.go [hunk 7]
```diff
@@ -39,7 +39,10 @@ import (
 // Pod permissions used to get OwnerReference corresponding to the current pod. This is required to ensure that
 // the operator pod is the leader in the given namespace. This will not be required if the leader election is done
 // by the manager, instead of the "leader" library
-//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
```
Comment (sebsoto) [inline cmd/operator/main.go:42]: why is pod watch being added?

File: bundle.Dockerfile [hunk 8]
```diff
@@ -36,10 +36,15 @@ LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
 LABEL operators.operatorframework.io.bundle.package.v1=windows-machine-config-operator
 LABEL operators.operatorframework.io.bundle.channels.v1=preview,stable
 LABEL operators.operatorframework.io.bundle.channel.default.v1=stable
+LABEL operators.operatorframework.io.metrics.builder=operator-sdk-v1.14.0+git
```
Comment (sebsoto) [inline bundle.Dockerfile:39]: why is this changed?

File: bundle/metadata/annotations.yaml [hunk 9]
```diff
@@ -6,10 +6,11 @@ annotations:
   operators.operatorframework.io.bundle.package.v1: windows-machine-config-operator
   operators.operatorframework.io.bundle.channels.v1: preview,stable
   operators.operatorframework.io.bundle.channel.default.v1: stable
+  operators.operatorframework.io.metrics.builder: operator-sdk-v1.14.0+git
   operators.operatorframework.io.metrics.mediatype.v1: metrics+v1
-  operators.operatorframework.io.metrics.builder: operator-sdk-v2.0.0+git
   operators.operatorframework.io.metrics.project_layout: go.kubebuilder.io/v3
   com.redhat.openshift.versions: "=v4.14"
+
```
Comment (sebsoto) [inline bundle/metadata/annotations.yaml:13]: Please revert these changes if unnecessary

File: Makefile [hunk 10]
```diff
@@ -173,6 +173,16 @@ OPM = $(shell which opm)
...
+USE_IMAGE_DIGESTS ?= false
+ifeq ($(USE_IMAGE_DIGESTS), true)
+    BUNDLE_GEN_FLAGS += --use-image-digests
+endif
```
Comment (jrvaldes) [inline Makefile:197]: Consider the `fi` keyword, as it is more commonly used and is considered the standard way to end an if statement block

File: Makefile [hunk 11]
```diff
@@ -108,6 +108,9 @@ build-daemon:
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run cmd/operator/main.go
 
+ifndef IGNORE-NOT-FOUND
+  IGNORE-NOT-FOUND = false
```
Comment (saifshaikh48) [inline Makefile:112]: nit: Constants usually are underscored `IGNORE_NOT_FOUND`
see top of makefile for examples

**Review by sebsoto — CHANGES_REQUESTED** [review]
Thanks @alinaryan PTAL at my comments

File: Makefile [hunk 12]
```diff
@@ -173,6 +173,16 @@ OPM = $(shell which opm)
...
 endif
 BUNDLE_IMGS ?= $(BUNDLE_IMG)
+# BUNDLE_GEN_FLAGS are the flags passed to the operator-sdk generate bundle command
+BUNDLE_GEN_FLAGS ?= -q --overwrite --version $(WMCO_VERSION) $(BUNDLE_METADATA_OPTS)
+
+# USE_IMAGE_DIGESTS defines if images are resolved via tags or digests
+# You can enable this value if you would like to use SHA Based Digests
+# To enable set flag to true
+USE_IMAGE_DIGESTS ?= false
+ifeq ($(USE_IMAGE_DIGESTS), true)
```
Comment (jrvaldes) [inline Makefile:195]: Nice usage of `ifeq`,  does `true`  needs to be surrounded by `"` given it is a string?

File: Makefile [hunk 13]
```diff
@@ -116,7 +119,7 @@ install: manifests kustomize ## Install CRDs into the K8s cluster specified in ~
 
 .PHONY: uninstall
 uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified in ~/.kube/config.
-	$(KUSTOMIZE) build config/crd | oc delete -f -
+	$(KUSTOMIZE) build config/crd | oc delete --IGNORE-NOT-FOUND=$(IGNORE-NOT-FOUND) -f -
```
Comment (saifshaikh48) [inline Makefile:122]: The flag should be `--ignore-not-found` otherwise i'm not sure oc will recognize it as a valid option

Comment (jrvaldes) [inline Makefile:122]: the flag name is `--ignore-not-found`, there may be a case mismatch here.

File: Makefile (hunk 11 above)
Comment (jrvaldes) [inline Makefile:112]: For the variable name, consider using snake case ` IGNORE_NOT_FOUND`

File: bundle.Dockerfile (hunk 8 above)
Comment (saifshaikh48) [reply bundle.Dockerfile:39]: in any case, shouldn't it be v1.32.0?

File: Makefile [hunk 14]
```diff
@@ -108,6 +108,9 @@ build-daemon:
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run cmd/operator/main.go
 
+ifndef IGNORE-NOT-FOUND
```
Comment (jrvaldes) [inline Makefile:111]: I feel the name of the variable lacks of context in the scope of the script; consider `IGNORE_NOT_FOUND_ON_DELETION`

File: bundle/metadata/annotations.yaml [hunk 15]
```diff
@@ -6,10 +6,11 @@ annotations:
   operators.operatorframework.io.bundle.package.v1: windows-machine-config-operator
   operators.operatorframework.io.bundle.channels.v1: preview,stable
   operators.operatorframework.io.bundle.channel.default.v1: stable
+  operators.operatorframework.io.metrics.builder: operator-sdk-v1.14.0+git
```
Comment (saifshaikh48) [inline bundle/metadata/annotations.yaml:9]: same as above, shouldn't this match the version from the CSV 1.32.0?

**Review by jrvaldes — COMMENTED** [review]
@alinaryan  Mostly LGTM.  PTAL at my comments.

File: Makefile [hunk 16]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
 CONTROLLER_GEN = $(shell pwd)/bin/controller-gen
 controller-gen: ## Download controller-gen locally if necessary.
 	$(call go-install-tool,$(CONTROLLER_GEN),sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.1)
+##@ Build Dependencies
 
+## Location to install dependencies to
+LOCALBIN ?= $(shell pwd)/bin
+$(LOCALBIN):
+	mkdir -p $(LOCALBIN)
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply Makefile:142]: if this is the code generated when using a newer version of operator-sdk it is fine leaving it as is. applies everywhere

File: Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:122]: thanks for catching

File: Makefile (hunk 12 above)
Comment (alinaryan) [reply Makefile:195]: I'd prefer to keep it to what is recommended by the OSDK docs. See https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.18.0/#support-image-digests-instead-of-tags

File: Makefile (hunk 10 above)
Comment (alinaryan) [reply Makefile:197]: Same as above

File: Makefile [hunk 17]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
...
+
+## Tool Binaries
+KUSTOMIZE ?= $(LOCALBIN)/kustomize
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:145]: Keeping this since it's OSDK generated code.

File: Makefile [hunk 18]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
...
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
+ENVTEST ?= $(LOCALBIN)/setup-envtest
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:147]: Keeping this since it's OSDK generated code.

File: Makefile [hunk 19]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
...
+
+## Tool Versions
+KUSTOMIZE_VERSION ?= v4.5.5
+CONTROLLER_TOOLS_VERSION ?= v0.10.0
+
+KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:153]: Keeping this since it's OSDK generated code.

File: Makefile [hunk 20]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
...
+$(KUSTOMIZE): $(LOCALBIN)
+	test -s $(LOCALBIN)/kustomize || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
+
+.PHONY: controller-gen
+controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
+$(CONTROLLER_GEN): $(LOCALBIN)
+	test -s $(LOCALBIN)/controller-gen || GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+
+.PHONY: envtest
+envtest: $(ENVTEST) ## Download envtest-setup locally if necessary.
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:165]: Keeping this since it's OSDK generated code.

File: Makefile [hunk 21]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
...
+$(ENVTEST): $(LOCALBIN)
+	test -s $(LOCALBIN)/setup-envtest || GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:167]: Keeping this since it's OSDK generated code.

File: Makefile (hunk 18 above)
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:147]: Please see the OSDK links in the commit message.

File: Makefile [hunk 22]
```diff
@@ -134,25 +134,37 @@ undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/confi
...
+	mkdir -p $(LOCALBIN)
+
+## Tool Binaries
+KUSTOMIZE ?= $(LOCALBIN)/kustomize
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
+ENVTEST ?= $(LOCALBIN)/setup-envtest
+
+## Tool Versions
+KUSTOMIZE_VERSION ?= v4.5.5
+CONTROLLER_TOOLS_VERSION ?= v0.10.0
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:151]: I would prefer to keep them where they are as it will make future OSDK upgrades easier to handle.

File: config/default/manager_auth_proxy_patch.yaml [hunk 23]
```diff
@@ -10,7 +10,7 @@ spec:
     spec:
       containers:
       - name: kube-rbac-proxy
-        image: gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
+        image: gcr.io/kubebuilder/kube-rbac-proxy:v0.13.0
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply config/default/manager_auth_proxy_patch.yaml]: I'd like to keep this version to stay as close to the recommended OSDK changes as possible. There might be a particular reason this version is being used. If we have a specific reason to use latest kube-rbac-proxy then it can be changed.

File: Makefile (hunk 14 above)
Comment (alinaryan) [reply Makefile:111]: I was going to change it to this, but I decided against it since this is OSDK generated code.

File: Makefile (hunk 11 above)
Comment (alinaryan) [reply Makefile:112]: I'd like to keep this as close to OSDK generated code as much as possible.

Comment (alinaryan) [reply Makefile:112]: I'd like to keep this as close to OSDK generated code as much as possible.

File: Makefile [hunk 24]
```diff
@@ -173,6 +173,16 @@ OPM = $(shell which opm)
 endif
 endif
 BUNDLE_IMGS ?= $(BUNDLE_IMG)
+# BUNDLE_GEN_FLAGS are the flags passed to the operator-sdk generate bundle command
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply Makefile:188]: Fixed

File: hack/pre-release.sh [hunk 25]
```diff
@@ -11,7 +11,7 @@ CVP=false
 BASEIMAGE=false
 DISTGIT=false
 GITLAB=false
-OPERATOR_SDK_VERSION=v1.15.0
+OPERATOR_SDK_VERSION=v1.32.0
```
[the remark this answers is not in the document]
Comment (alinaryan) [reply hack/pre-release.sh:14]: Can tackle this in a followup PR

File: hack/pre-release.sh [hunk 26]
```diff
@@ -227,7 +227,7 @@ github_update() {
 
   # remove make bundle artifacts
   sed -i 's/REPLACE_IMAGE:latest/REPLACE_IMAGE/' bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
-  sed -i 's/operator-sdk-v1.14.0+git/operator-sdk-v1.15.0+git/' bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
+  sed -i "s/operator-sdk-v1.14.0+git/operator-sdk-$OPERATOR_SDK_VERSION+git/" bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
```
Comment (saifshaikh48) [inline hack/pre-release.sh:230]: This sed command can likely get removed entirely now. Fine with that happening in a separate PR though

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

File: hack/pre-release.sh (hunk 26 above)
Comment (alinaryan) [reply hack/pre-release.sh:230]: Can tackle in followup PR
